llvm/tools/llvm-reduce/deltas/ReduceValuesToReturn.cpp

```diff
@@ -0,0 +1,244 @@
+    VMap[&OldArg] = &NewArg;          // Add mapping to VMap
+  }
+
+  SmallVector<ReturnInst *, 8> Returns; // Ignore returns cloned.
+  CloneFunctionInto(NewF, &OldF, VMap,
+                    CloneFunctionChangeType::LocalChangesOnly, Returns, "",
+                    /*CodeInfo=*/nullptr);
+  OldF.replaceAllUsesWith(NewF);
+  OldF.eraseFromParent();
+}
+
+// Check if all the callsites of the void function are void, or happen to
+// incorrectly use the new return type.
+//
+// TODO: We could make better effort to handle call type mismatches.
+static bool canReplaceFuncUsers(const Function &F, Type *NewRetTy) {
+  for (const Use &U : F.uses()) {
+    const CallBase *CB = dyn_cast<CallBase>(U.getUser());
+    if (!CB)
+      continue;
+
+    // Normal pointer uses are trivially replacable.
+    if (!CB->isCallee(&U))
+      continue;
+
+    // We can trivially replace the correct void call sites.
+    if (CB->getType()->isVoidTy())
+      continue;
+
+    // We can trivially replace the call if the return type happened to match
+    // the new return type.
+    if (CB->getType() == NewRetTy)
+      continue;
+
+    LLVM_DEBUG(dbgs() << "Cannot replace callsite with wrong type: " << *CB
+                      << '\n');
+    return false;
+  }
+
+  return true;
+}
+
+/// Return true if it's worthwhile replacing the non-void return value of \p BB
+/// with \p Replacement
+static bool shouldReplaceNonVoidReturnValue(const BasicBlock &BB,
+                                            const Value *Replacement) {
+  if (const auto *RI = dyn_cast<ReturnInst>(BB.getTerminator()))
+    return RI->getReturnValue() != Replacement;
+  return true;
+}
+
+static bool canHandleSuccessors(const BasicBlock &BB) {
+  // TODO: Handle invoke and other exotic terminators
+  if (!isa<ReturnInst, UnreachableInst, BranchInst, SwitchInst>(
+          BB.getTerminator()))
+    return false;
+
+  for (const BasicBlock *Succ : successors(&BB)) {
+    if (!Succ->canSplitPredecessors())
+      return false;
+  }
+
+  return true;
+}
+
+static bool shouldForwardValueToReturn(const BasicBlock &BB, const Value *V,
+                                       Type *RetTy) {
+  if (!isReallyValidReturnType(V->getType()))
+    return false;
+
+  return (RetTy->isVoidTy() ||
+          (RetTy == V->getType() && shouldReplaceNonVoidReturnValue(BB, V))) &&
+         canReplaceFuncUsers(*BB.getParent(), V->getType());
+}
+
+static bool tryForwardingInstructionsToReturn(
+    Function &F, Oracle &O,
+    std::vector<std::pair<Function *, Value *>> &FuncsToReplace) {
+
+  // TODO: Should we try to expand returns to aggregate for function that
+  // already have a return value?
+  Type *RetTy = F.getReturnType();
+
+  for (BasicBlock &BB : F) {
+    if (!canHandleSuccessors(BB))
+      continue;
+
+    for (Instruction &I : BB) {
+      if (shouldForwardValueToReturn(BB, &I, RetTy) && !O.shouldKeep()) {
+        FuncsToReplace.emplace_back(&F, &I);
+        return true;
```

Comment:
Hm, does the oracle properly handle this kind of early return?

Basically if O.shouldKeep() return false and now return true, we'll now perform an extra query (for the next instruction in the block), effectively shifting all the indices?

Reply:
It kind of has to, or none of the reductions that have side effects on other code would work. The chunks get recounted after each step